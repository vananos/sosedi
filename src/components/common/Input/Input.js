import React, { Component } from "react";
import infoCircle from "../../../assets/common/info-circle-solid.svg";
import PropTypes from "prop-types";
import "./Input.scss";

export default class Input extends Component {
  constructor(props) {
    super(props);
    this.state = {
      value: props.value || "",
      changeHandler: props.onChange
    };
    this.inputRef = React.createRef();
  }

  handleChange = e => {
    this.state.changeHandler(e, this);
    const newValue = e.target.value;
    this.setState({
      value: newValue
    });
  };

  setValue = value => {
    this.setState({ value });
  };

  render() {
    const {
      type,
      name,
      id,
      label,
      info,
      max,
      className,
      onBlur,
      error: { error: errorMsg, value: wrongValue } = {}
    } = this.props;

    const currentValue = this.state.value;

    const mustShowError = errorMsg && this.state.value === wrongValue;
    return (
      <div
        className={`input-wrapper ${mustShowError ? "error" : ""} ${className}`}
        onClick={e => this.inputRef.current.focus()}
      >
        <input
          ref={this.inputRef}
          type={type}
          className="input"
          name={name}
          id={id}
          value={currentValue}
          onChange={this.handleChange}
          onBlur={onBlur}
          max={max}
        />
        <div className={`label-info ${currentValue !== "" ? "filled" : ""}`}>
          <span className="label">{label}</span>
          {info && (
            <img src={infoCircle} alt="info" width="10px" height="10px" />
          )}
        </div>
        {mustShowError && <span className="error-description">{errorMsg}</span>}
      </div>
    );
  }
}

Input.defaultProps = {
  type: "text",
  className: "",
  onChange: () => {}
};

Input.propTypes = {
  type: PropTypes.string,
  name: PropTypes.string,
  value: PropTypes.string,
  id: PropTypes.string,
  label: PropTypes.string.isRequired,
  info: PropTypes.node,
  className: PropTypes.string,
  error: PropTypes.object,
  onChange: PropTypes.func,
  onBlur: PropTypes.func
};
