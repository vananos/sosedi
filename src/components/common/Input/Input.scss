@import "../../../base.scss";

.input {
    display: block;
    border: none;
    outline: none;
    padding: 10 0;
    font-size: 1em;
    border-bottom: 1px solid $light-grey;
    width: 100%;

    &-wrapper {
        display: block;
        position: relative;
        padding: 0;
        padding-top: 1em;
        width: 100%;
    }

    &:focus {
        border-color: $brown;
    }
}

.error .input {
    font-size: 1em;
    border-bottom-color: $red !important;
}

.error-description {
    width: 100%;
    color: $red;
    font-size: 0.8em;
}

.label-info {
    position: absolute;
    left: 0;
    top: 1.25em;
    transition: 0.3s linear;
    text-transform: capitalize;
    color: $light-grey;
}

.input:focus~.label-info,
.filled {
    top: 0;
    right: 0;
    left: auto;
    font-size: 0.75em;
    color: rgba(222, 170, 93, 1);
}

.error .input:focus~.label-info,
.error .filled {
    color: $red;
}