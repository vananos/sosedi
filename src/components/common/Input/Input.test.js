import React from "react";
import Input from "./Input";
import { render, mount } from "enzyme";

describe("Input, ", () => {
  const defaultInputProps = () => ({
    value: "",
    name: "testInput",
    id: "testInputId",
    type: "text",
    label: "test text",
    className: "customClass"
  });

  it("when value is empty -> label is on bottom", () => {
    const inputWrapper = render(<Input {...defaultInputProps()} />);

    expect(inputWrapper).toMatchSnapshot();
  });

  it("when value provided -> label is on top", () => {
    const inputProps = defaultInputProps();
    inputProps.value = "value is provided";
    const inputWrapper = render(<Input {...inputProps} />);

    expect(inputWrapper).toMatchSnapshot();
  });

  it("when focused -> label is on top", () => {
    const inputWrapper = mount(<Input {...defaultInputProps()} />);
    inputWrapper
      .find("input")
      .first()
      .simulate("focus");

    expect(inputWrapper).toMatchSnapshot();
  });

  it("when error is provideed -> should see error message", () => {
    const error = {
      error: "it's definitely wrong",
      value: "wrong value"
    };
    const inputProps = defaultInputProps();
    inputProps.error = error;

    const inputWrapper = render(<Input {...inputProps} />);

    expect(inputWrapper).toMatchSnapshot();
  });
});
