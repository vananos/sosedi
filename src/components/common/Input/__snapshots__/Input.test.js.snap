// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input,  when error is provideed -> should see error message 1`] = `
<div
  class="input-wrapper  customClass"
>
  <input
    class="input"
    id="testInputId"
    name="testInput"
    type="text"
    value=""
  />
  <div
    class="label-info "
  >
    <span
      class="label"
    >
      test text
    </span>
  </div>
</div>
`;

exports[`Input,  when focused -> label is on top 1`] = `
<Input
  className="customClass"
  id="testInputId"
  label="test text"
  name="testInput"
  type="text"
  value=""
>
  <div
    className="input-wrapper  customClass"
    onClick={[Function]}
  >
    <input
      className="input"
      id="testInputId"
      name="testInput"
      onChange={[Function]}
      type="text"
      value=""
    />
    <div
      className="label-info "
    >
      <span
        className="label"
      >
        test text
      </span>
    </div>
  </div>
</Input>
`;

exports[`Input,  when value is empty -> label is on bottom 1`] = `
<div
  class="input-wrapper  customClass"
>
  <input
    class="input"
    id="testInputId"
    name="testInput"
    type="text"
    value=""
  />
  <div
    class="label-info "
  >
    <span
      class="label"
    >
      test text
    </span>
  </div>
</div>
`;

exports[`Input,  when value provided -> label is on top 1`] = `
<div
  class="input-wrapper  customClass"
>
  <input
    class="input"
    id="testInputId"
    name="testInput"
    type="text"
    value="value is provided"
  />
  <div
    class="label-info filled"
  >
    <span
      class="label"
    >
      test text
    </span>
  </div>
</div>
`;
